VSRAD.Package/DebugVisualizer/Wavemap/WavemapCanvas.cs

```diff
@@ -0,0 +1,137 @@
+            {
+                foreach (var row in _rectangles)
+                    foreach (var el in row)
+                        el.Visibility = System.Windows.Visibility.Hidden;
+                return;
+            }
+
+            if (view.WavesPerGroup != _rectangles.Count)
+            {
+                if (view.WavesPerGroup > _rectangles.Count)
+                {
+                    var rCount = _rectangles.Count;
+                    for (int i = rCount; i < view.WavesPerGroup; ++i)
+                    {
+                        _rectangles.Add(new List<Rectangle>());
+                    }
+                }
+                else
+                {
+                    for (int i = view.WavesPerGroup; i < _rectangles.Count; ++i)
+                        for (int j = 0; j < _rectangles[i].Count; ++j)
+                            _rectangles[i][j].Visibility = System.Windows.Visibility.Hidden;
+                }
+            }
+
+            for (int j = 0; j < _rectangles.Count; ++j)
+            {
+                if (_rectangles[j].Count != view.GroupCount)
+                {
+                    if (view.GroupCount > _rectangles[j].Count)
+                    {
+                        for (int i = _rectangles[j].Count; i < view.GroupCount; ++i)
+                        {
+                            var r = InitiateWaveRectangle(j, i);
+                            _rectangles[j].Add(r);
+                            _canvas.Children.Add(r);
+                        }
+                    }
+                    else
+                    {
+                        for (int i = view.GroupCount; i < _rectangles[j].Count; ++i)
+                            _rectangles[j][i].Visibility = System.Windows.Visibility.Hidden;
+                    }
+                }
+            }
+
+            for (int i = 0; i < view.GroupCount; ++i)
+                for (int j = 0; j < view.WavesPerGroup; ++j)
+                    UpdateWaveRectangle(j, i);
+            _canvas.InvalidateVisual();
+        }
+
+        private Rectangle InitiateWaveRectangle(int row, int column)
+        {
+            var r = new Rectangle();
+            r.Visibility = System.Windows.Visibility.Hidden;
+            r.Height = _rectangleSize;
+            r.Width = _rectangleSize;
+            r.StrokeThickness = 1;
+            r.Stroke = Brushes.Black;
+            Canvas.SetLeft(r, 1 + (_rectangleSize - 1) * column);
+            Canvas.SetTop(r, 1 + (_rectangleSize - 1) * row);
+            return r;
+        }
+
+        private void UpdateWaveRectangle(int row, int column)
+        {
+            var r = _rectangles[row][column];
+            var validWave = _view != null && _view.IsValidWave(row, column);
+            r.ToolTip = new ToolTip()
+            {
+                Content = validWave ?
+                    $"Group: {_view[row, column].GroupIdx}\nWave: {_view[row, column].WaveIdx}\nLine: {_view[row, column].BreakLine}"
```

Comment:
It's better to assign `_view[row, column]` to a local variable instead of repeatedly calling the indexer, since it isn't a simple memory lookup, you have additional logic there.

Reply:
Thanks for suggestion, resolved in 2712ed9

Resolving this needed additional refactoring, since `[row, column]` getter did not check any boundaries, so if I just rewrote it as is, we could get out of bounds exception while trying to get break line of hidden group.

So, I generally improved `WavemapView` interface by making `IsValidWave` private, check it in getter and return default `WaveInfo` struct if requested wave is out of bounds of given data.